src/libs/FileSorter.cpp

```diff
@@ -0,0 +1,193 @@
+#include "FileStream.h"
+
+#include <new>
+#include <malloc.h>
+#include <string.h>
+
+
+FileSorter::FileSorter(string dir_path, __file_sort_filter_fn_t callback)
+{
+    this->directory_path = dir_path;
+    this->error = false;
+    this->file_info_array = NULL;
+    this->file_count = 0;
+    this->total_file_count = 0;
+    this->file_index = -1;
+    this->filter_callback = callback;
+
+    open_directory(dir_path);
+}
+
+FileSorter::~FileSorter()
+{
+    delete_file_info_array();
+}
+
+int FileSorter::compare_dir(file_info_t* file_a, file_info_t* file_b)
+{
+    return strcmp(file_a->file_name, file_b->file_name);
+}
+
+size_t FileSorter::count_files_in_folder(string dir_path)
+{
+    struct dirent *p;
+    uint16_t count = 0;
+    DIR* folder = opendir(dir_path.c_str());
+    if ( folder != NULL ) {
+        while ( (p = readdir(folder)) != NULL ) {
+            if ( (this->filter_callback == NULL) || this->filter_callback(p) ) {
+                count++;
+            }
+        }
+        closedir(folder);
+        return count;
+    }
+    return 0;
+}
+
+void FileSorter::sort_directory(void)
+{
+    // sort the array.
+    qsort(this->file_info_array, this->file_count, sizeof(file_info_t), (__compar_fn_t)compare_dir);
+    this->file_index = -1;
+}
+
+void FileSorter::delete_file_info_array(void)
+{
+    if ( this->file_info_array != NULL) {
+        // file names were allocated previously with strdup - free them now.
+        for ( size_t i=0; i<this->file_count; i++ ) {
+            free(this->file_info_array[i].file_name);
+        }
+
+        // free the array.
+        delete[] this->file_info_array;
+    }
+}
+
+void FileSorter::open_directory(string dir_path)
+{
+    // reset all member state variables.
+    this->directory_path = dir_path;
+    this->error = false;
+    this->file_count = 0;
+    this->total_file_count = 0;
+    this->file_index = -1;
+    delete_file_info_array();
+
+    DIR *d;
+    d = opendir(this->directory_path.c_str());
+    if ( !(this->error = (d == NULL)) ) {
+        // get a count of items in the folder.
+        this->total_file_count = count_files_in_folder(this->directory_path);
+
+        // allocate the array used for sorting.
+        this->file_info_array = new (std::nothrow) file_info_t[this->total_file_count];
+
+        // read the file info into the array.
+        if ( !(this->error = (this->file_info_array == NULL)) ) {
+            struct dirent* file_info;
+            file_info_t* array_entry;
+            this->file_count = 0;
+            array_entry = this->file_info_array;
+            while ( this->file_count < this->total_file_count && (file_info = readdir(d)) != NULL ) {
+                if ( (this->filter_callback == NULL) || this->filter_callback(file_info) ) {
+                    array_entry->is_dir = file_info->d_isdir;
+                    array_entry->file_size = file_info->d_fsize;
+                    array_entry->file_name = strdup(file_info->d_name);
```

Comment:
strdup can fail, so you need to test for nullptr here, and then clean up if it fails.